cobra/util/solver.py

```diff
@@ -26,6 +29,108 @@ class SolverNotFound(Exception):
+    """ Set the model objective
+
+    Parameters
+    ----------
+    model: cobra model
+       The model to set the objective for
+    value: model.solver.interface.Objective such as
+           optlang.glpk_interface.Objective, sympy.Basic, dict, int,
+           list, string or Reaction
+
+        If the model objective is linear, the value can be a new Objective
+        object directly, a sympy expression which is used to create a new
+        Objective, a dictionary of linear coefficients where each key is a
+        reaction and the element the new coefficient (float), a list of
+        reactions where each reaction is interpreted to be meant to have an
+        objective coefficient of 1, a string interpreted as a single
+        reaction identifier or directly a Reaction object.
+
+        If the objective is not linear, only values of class Objective or
+        sympy expression's are allowed.
+    additive: bool
+        If true, add the terms to the current objective, otherwise start with
+        an empty objective
+    """
+    by_expression = (isinstance(value, sympy.Basic) or
+                     isinstance(value, model.solver.interface.Objective))
+    by_coefficient = isinstance(value, (dict, int, list, six.string_types,
+                                        cobra.Reaction))
+    not_supported = (
+        additive and (not model.objective.is_Linear and not by_expression))
+    if not_supported:
+        raise ValueError('can only update non-linear objectives additively '
+                         'using object of class '
+                         'model.solver.interface.Objective, not %s' %
+                         type(value))
+    if by_expression:
+        if not additive:
+            if isinstance(value, sympy.Basic):
+                model.solver.objective = model.interface.Objective(
+                    value, sloppy=False)
+            if isinstance(value, model.solver.interface.Objective):
+                model.solver.objective = value
+        else:
+            if isinstance(value, model.solver.interface.Objective):
+                value = value.expression
+            model.solver.objective += value
+            value = -value
+    elif by_coefficient:
+        if not additive:
+            model.solver.objective = model.solver.interface.Objective(
+                sympy.S.Zero, direction='max')
+        if not isinstance(value, dict):
+            value = {rxn: 1 for rxn in model.reactions.get_by_any(value)}
+        for reaction, coef in value.items():
+            value[reaction] = reaction.objective_coefficient
+            model.solver.objective.set_linear_coefficients(
+                {reaction.forward_variable: coef,
+                 reaction.reverse_variable: -coef})
+    else:
+        raise TypeError(
+            '%r is not a valid objective for %r.' % (value, model.solver))
+    context = get_context(model)
+    if context:
+        context(partial(set_objective, model=model, value=value,
```

Comment:
Will this work? I guess I'm confused by how `value` becomes the old value we want to reset to. It might be clearer if we could use a new variable (`old_value`?) to keep track of what should be undone with the context manager.